acceptancetests/deploy_stack.py

```diff
@@ -570,6 +571,40 @@ def tear_down(self, has_controller):
             self.tear_down_client.kill_controller(check=True)
 
 
+class ExistingController:
+    """A Controller strategy where the controller is already present.
+
+    Intended for use with BootstrapManager.
+    """
+
+    def __init__(self, client, tear_down_client, juju_data_dir, env):
+        self.client = client
+        self.client.env.environment = env
+        self.tear_down_client = tear_down_client
+
+    def create_initial_model(self, upload_tools):
+        """Create the initial model."""
+        self.client.add_model(self.client.env)
+        logging.info('Added model {} to existing controller'.format(
+            self.client.env.environment))
+
+    def prepare(self):
+        """Prepare client for use by killing the existing controller."""
+        self.tear_down_client.kill_controller()
+
+    def get_hosts(self):
+        """Provide the controller host."""
+        host = get_machine_dns_name(
+            self.client.get_controller_client(), '0')
+        if host is None:
+            raise ValueError('Could not get machine 0 host')
+        return {'0': host}
+
+    def tear_down(self, _):
+        """Model already destroyed, switch back to a known default"""
+        self.client.switch('controller')
```

Comment:
is there a need to switch? There is nowhere else in the jujupy codebase that expects a model to be selected (and expects commands to provide a fully qualified model/controller arg.)

The expectation is that the test calling using this controller strategy cleans up the model that's created?
It would be cleaner to have that cleanup happen here. Otherwise there is a burden on the caller to know that they have some cleanup they need to do (including when an exception occurs etc.) which complicates things (and muddies the clear line of responsibilities).